### PR TITLE
[SPARK-2208] Fix for local metrics tests can fail on fast machines

### PR DESCRIPTION
**Pull request by joan38**
## What changes were proposed in this pull request?

A fix for local metrics tests that can fail on fast machines.
This is probably what is suggested here #3380 by @aarondav?
## How was this patch tested?

CI Tests

Cheers

### CODE REVIEW

File: core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala [hunk 1]
```diff
@@ -19,14 +19,17 @@ package org.apache.spark.scheduler
 
 import java.util.concurrent.Semaphore
 
-import scala.collection.mutable
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: These imports were in the right place; restore them

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala [hunk 2]
```diff
@@ -27,12 +27,12 @@ trait LocalSparkContext extends BeforeAndAfterEach with BeforeAndAfterAll { self
 
   @transient var sc: SparkContext = _
 
-  override def beforeAll() {
+  override def beforeAll(): Unit = {
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/LocalSparkContext.scala]: Although this change is not necessary or related, it's not wrong

File: core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala [hunk 3]
```diff
@@ -476,3 +480,19 @@ private class ListenerThatAcceptsSparkConf(conf: SparkConf) extends SparkListene
   var count = 0
   override def onJobEnd(job: SparkListenerJobEnd): Unit = count += 1
 }
+
+private class SlowShuffleManager(conf: SparkConf) extends SortShuffleManager(conf) {
+  
+  override val shuffleBlockResolver: IndexShuffleBlockResolver =
+    new SlowIndexShuffleBlockResolver(conf)
+    
+    private class SlowIndexShuffleBlockResolver(conf: SparkConf,
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: (Can you just make this an anonymous inner class? might be shorter. I might also be missing something.)

File: core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala [hunk 4]
```diff
@@ -476,3 +480,19 @@ private class ListenerThatAcceptsSparkConf(conf: SparkConf) extends SparkListene
...
+                                                _blockManager: BlockManager = null)
+      extends IndexShuffleBlockResolver(conf, _blockManager) {
+
+      override def getBlockData(blockId: ShuffleBlockId): ManagedBuffer = {
+        Thread.sleep(1)
+        super.getBlockData(blockId)
+      }
+    }
+}
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: I forget whether this will fail style checks without a newline at the end.

Content-wise, this looks good though

File: core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala (hunk 1 above)
Comment (joan38) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: Sorry it was Idea's default imports reorganiser, I will apply the setting shown here:
https://cwiki.apache.org/confluence/display/SPARK/Spark+Code+Style+Guide#SparkCodeStyleGuide-Imports

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala (hunk 2 above)
Comment (joan38) [inline core/src/test/scala/org/apache/spark/LocalSparkContext.scala]: The procedure notation is deprecated.
Should I rollback this change?

File: core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala (hunk 3 above)
Comment (joan38) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: Would you prefer:
```
private class SlowShuffleManager(conf: SparkConf) extends SortShuffleManager(conf) {

  override val shuffleBlockResolver: IndexShuffleBlockResolver =
    new IndexShuffleBlockResolver(conf) {

      override def getBlockData(blockId: ShuffleBlockId): ManagedBuffer = {
        Thread.sleep(1)
        super.getBlockData(blockId)
      }
    }
}
```

Comment (srowen) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: Exactly, that seems simpler

File: core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala (hunk 4 above)
Comment (joan38) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: I added a new line

File: core/src/test/scala/org/apache/spark/LocalSparkContext.scala (hunk 2 above)
Comment (srowen) [inline core/src/test/scala/org/apache/spark/LocalSparkContext.scala]: I think you could take it out of this PR. There are so many instances of this syntax that I'm not even sure it's worth changing everywhere. It's fine to fix if you're already modifying the surrounding code, but this is unrelated

**Comment by srowen** [comment]
Jenkins test this please

**Comment by srowen** [comment]
Darn, looks like the test still fails

**Comment by joan38** [comment]
Yeah I'm investigating but any ideas are welcome.

**Comment by srowen** [comment]
@joan38 how about a longer wait? 1ms may not be enough. Did you mean 1 second?

**Comment by joan38** [comment]
I tried with 1000ms and the test runs forever.
Should we try 10ms? 100ms works too but the test runs for a minute...

**Comment by srowen** [comment]
OK, makes sense. Yeah, whatever makes it pass comfortably. 10ms?

**Comment by joan38** [comment]
Can we try to run Jenkins? I did the change for 10ms.
It anyway always pass on my PC as it's a slow machine.

File: core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala [hunk 5]
```diff
@@ -255,21 +262,19 @@ class SparkListenerSuite extends SparkFunSuite with LocalSparkContext with Match
...
       if (stageInfo.rddInfos.exists(_.name == d4.name)) {
         checkNonZeroAvg(
           taskInfoMetrics.map(_._2.shuffleReadMetrics.get.fetchWaitTime),
           stageInfo + " fetchWaitTime")
       }
-      */
 
       taskInfoMetrics.foreach { case (taskInfo, taskMetrics) =>
-        taskMetrics.resultSize should be > (0L)
+        taskMetrics.resultSize should be > 0L
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: I'm not a scalatest expert but I seem to recall you need these parens. We'll see what Jenkins says.

**Comment by srowen** [comment]
Jenkins retest this please

File: core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala (hunk 5 above)
Comment (joan38) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: This:
taskMetrics.resultSize should be > 0L
 is analogous to:
taskMetrics.resultSize.should(be.>(0L))

**Comment by joan38** [comment]
Damn, it failed again! Should we try with 100ms?
I just pushed the 100ms version.

**Comment by srowen** [comment]
Jenkins retest this please

File: core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala [hunk 6]
```diff
@@ -255,21 +262,19 @@ class SparkListenerSuite extends SparkFunSuite with LocalSparkContext with Match
...
         if (stageInfo.rddInfos.exists(info => info.name == d2.name || info.name == d3.name)) {
-          taskMetrics.inputMetrics should not be ('defined)
-          taskMetrics.outputMetrics should not be ('defined)
+          taskMetrics.inputMetrics should not be 'defined
+          taskMetrics.outputMetrics should not be 'defined
           taskMetrics.shuffleWriteMetrics should be ('defined)
-          taskMetrics.shuffleWriteMetrics.get.bytesWritten should be > (0L)
+          taskMetrics.shuffleWriteMetrics.get.bytesWritten should be > 0L
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: These changes don't seem to matter, yeah, but they also don't do anything and these lines are now inconsistent. They could just be reverted.

So the test passes but now takes 52 seconds. I wonder if something in the middle is a good compromise. 25ms?

Comment (joan38) [inline core/src/test/scala/org/apache/spark/scheduler/SparkListenerSuite.scala]: Ah yeah, that's because I usually write:
taskMetrics.shuffleWriteMetrics shouldBe 'defined
instead of:
taskMetrics.shuffleWriteMetrics should be ('defined)
I find it much quicker to write and Idea doesn't warn me on the useless brackets. Never mind I reverted this.

I just pushed the 25ms version. "Jenkins retest this please" :stuck_out_tongue_winking_eye:

**Comment by srowen** [comment]
Jenkins add to whitelist

**Comment by srowen** [comment]
Jenkins retest this please

**Comment by joan38** [comment]
Failed for 10ms, 25ms, 50ms. So trying with 75ms as it passes for 100ms.

**Comment by srowen** [comment]
Hm, tough one. It sounds like it needs to be at least 100ms, but I wonder if that's even reliable. If it's not working after that, maybe we just drop this test.

**Comment by joan38** [comment]
I'm thinking that maybe this:
```scala
      // just to make sure some of the tasks take a noticeable amount of time 
      val w = { i: Int => 
        if (i == 0) { 
          Thread.sleep(100) 
        }
      }
```
Is our issue.
So I want to try with the `Thread.sleep(100)` commented.

**Comment by joan38** [comment]
Ahah, I knew it!
I will tidy up this and push again with only 10ms.

**Comment by joan38** [comment]
All good, this is ready to merge I guess.

**Comment by srowen** [comment]
OK so this takes out some injected delay, and now relies only on the delay induced by the shuffle manager. @squito does that sound right? I think you wrote the key block being removed here. It passes, it seems, but want to make sure we're not removing the point of the test. It seems like it's still testing what it intends.

**Comment by srowen** [comment]
Merged to master. I'm fairly confident that the change doesn't change the original intent.

**Comment by JoshRosen** [comment]
The local metrics suite tests are failing more frequently now; is it possible that this patch could be related? See https://spark-tests.appspot.com/tests/org.apache.spark.scheduler.SparkListenerSuite/local%20metrics

**Comment by joan38** [comment]
Yes, it looks like our change did not resolve this issue.
Maybe we should go for more than just 10ms. 100ms?

**Comment by srowen** [comment]
Darn, still flaky then. It's either give up and revert or try a longer value. Although it did seem it was just a question of rewriting the test to put enough delay in the right place, I wonder if that's really the issue. Well, I suppose I favor trying to fix forward once before giving up on this text permanently. You can make a PR for this same JIRA

**Comment by JoshRosen** [comment]
I think this patch made things worse on Jenkins because prior to this patch we didn't observe a master builder failure for an entire week, whereas now it's failing over half the time. I'm in favor of reverting for now.

**Comment by srowen** [comment]
Yeah let's revert it then. I'm not sure why the behavior is so different. Eh, this is a dumb question, but I'm not sure I've done a revert the normal way in Spark before. Is it different from just git revert and pushing? any special script or practice.

**Comment by JoshRosen** [comment]
Yep, just `git revert` then push. When I do a revert I'll usually make sure to comment on the original PR and I'll also re-open any JIRAs that the reverted PR was supposed to fix.

**Comment by joan38** [comment]
Sorry for the mess I introduced.
I will make a new PR with 100ms once it's reverted and this time we might want to run multiple times Jenkins to be sure that it works.

**Comment by srowen** [comment]
(I reverted this one in master for now)

**Comment by andrewor14** [comment]
Well if the patch in master is bad I don't think the one in 1.6 will be any better.

**Comment by srowen** [comment]
Oh it was only ever merged in master. I didn't mean "in master" as opposed to not in other branches

**Comment by andrewor14** [comment]
oh I see, never mind then
